pkg/filesystem/subvolume.go

```diff
@@ -211,7 +217,89 @@ func Delete(ctx context.Context, clientsets *k8sutil.Clientsets, OperatorNamespa
+		exec.RunCommandInOperatorPod(ctx, clientsets, "rados", []string{"rm", omapdetails.omapval, "-p", poolName, "--namespace", "csi"}, OperatorNamespace, CephClusterNamespace, true)
+		logging.Info("omap object:%q deleted", omapdetails.omapval)
+	}
+	if omapdetails.omapkey != "" {
+		// remove omap key.
+		exec.RunCommandInOperatorPod(ctx, clientsets, "rados", []string{"rmomapkey", "csi.volumes.default", omapdetails.omapkey, "-p", poolName, "--namespace", "csi"}, OperatorNamespace, CephClusterNamespace, true)
+		logging.Info("omap key:%q deleted", omapdetails.omapkey)
+	}
+}
+
+// getOmapDetails gets the omap key and value details for a given subvolume.
+// Deletion of omap object required the rbd-image-name which is of format
+// csi.volume.subvolume, where subvolume is the name of subvolume that needs to be
+// deleted.
+// similarly to delete of omap key requires csi.volume.ompakey, where
+// omapkey is the pv name which is extracted the omap object.
+func getOmapDetails(ctx context.Context, clientsets *k8sutil.Clientsets, OperatorNamespace, CephClusterNamespace, subVol, fs string) omapDetails {
+
+	var details omapDetails
+	var omapval, pv string
+	poolName := fs + "-metadata"
+	replace := strings.Replace(subVol, "-", ".", 2)
+	checknfs := strings.Contains(subVol, "nfs-export")
+	// check if it is cephfs or nfs subvolume
+	if checknfs {
+		omapval = strings.Replace(replace, "export", "volume", 1)
+		omapval = strings.Replace(omapval, "nfs", "csi", 1)
+	} else {
+		omapval = strings.Replace(replace, "vol", "volume", 1)
+
+	}
+	ls, err := exec.RunCommandInOperatorPod(ctx, clientsets, "rados", []string{"listomapvals", omapval, "-p", poolName, "--namespace", "csi"}, OperatorNamespace, CephClusterNamespace, true)
+	if err != nil {
+		if !strings.Contains(err.Error(), "No such file or directory") {
+			logging.Fatal(fmt.Errorf("failed to list omapvals: %q", err))
+		} else {
+			logging.Info("No omapvals found for subvolume %s", subVol)
+			return omapDetails{}
+		}
+	}
+	if ls != "" {
+
+		// Below is how the output of listmapvals looks, and when we split it w.r.t
+		// '|' we can get the pv name at 7,9 and 11th part of the output.
+		// Below example is for cephfs, we have similar case for nfs subvolumes.
+		//csi.imagename
+		//value (44 bytes) :
+		//00000000  63 73 69 2d 76 6f 6c 2d  35 66 62 32 61 38 30 32  |csi-vol-5fb2a802|
+		//00000010  2d 38 37 31 31 2d 34 33  63 33 2d 62 39 36 35 2d  |-8711-43c3-b965-|
+		//00000020  66 66 37 65 37 39 35 62  33 35 33 65              |ff7e795b353e|
+		//0000002c
+		//csi.volname
+		//value (40 bytes) :
+		//00000000  70 76 63 2d 65 63 62 62  34 34 38 61 2d 30 35 37  |pvc-ecbb448a-057|
+		//00000010  36 2d 34 35 30 35 2d 62  32 39 35 2d 30 32 39 33  |6-4505-b295-0293|
+		//00000020  34 37 37 36 34 65 61 64                           |47764ead|
+		//0000002
+		volname := strings.Split(ls, "|")
+		if checknfs {
+			if len(volname) > 16 {
+				pv = "csi.volume." + volname[12] + volname[14] + volname[16]
```

Comment:
This looks very brittle. If anything changes in the output of the command or if the volume name could be a different length, this can easily break. Is there no way to get the csi.imagename without the binary output?

Reply:
I am not sure, I initially tried to get the value for csi.imagename only, will see if there is a way possible

Reply:
A much cleaner way is to write to a temporary file:
```console
sh-5.1$ rados --pool ocs-storagecluster-cephblockpool getomapval csi.volume.df42d7a3-8f3a-4630-835e-ed264a088777 csi.imagename /tmp/imagename 
Writing to /tmp/imagename
sh-5.1$ grep '' /tmp/imagename
csi-vol-df42d7a3-8f3a-4630-835e-ed264a088777
```
Note that the contents of `/tmp/imagename` does not contain a `\n` at the end of the line.

You can also write to `/dev/stdout` so that you do not need to create/remove a temporary file:
```console
sh-5.1$ IMAGENAME=$(rados --pool ocs-storagecluster-cephblockpool getomapval csi.volume.df42d7a3-8f3a-4630-835e-ed264a088777 csi.imagename /dev/stdout 2> /dev/null)
sh-5.1$ echo ${IMAGENAME}
csi-vol-df42d7a3-8f3a-4630-835e-ed264a088777
```

A big _**NO**_ against the parsing of the output the way you do now, sorry!

Reply:
Yeah Yeah., I am working on it. I was bit busy testing the ci tests, will work on other comments too.